src/compiler/checker.ts

```diff
@@ -13335,7 +13335,10 @@ namespace ts {
                                         //   type Foo<T extends string, U extends T> = [T, U];
                                         //   type Bar<T> = T extends Foo<infer X, infer X> ? Foo<X, X> : T;
                                         // the instantiated constraint for U is X, so we discard that inference.
-                                        const mapper = createTypeMapper(typeParameters, getEffectiveTypeArguments(typeReference, typeParameters));

```

Comment:
In the simple repro:

```ts
type AMappedType<T> = { [KeyType in keyof T]: number };

type HasM = {
  m: number;
};

// Simplified repro from #48059

interface X1<
  T extends HasM,
  Output = AMappedType<{ s: number; } & { [k in keyof T]: number; }>
> {
  tee: T;
  output: Output;
}

type F1<T> = T extends X1<infer U> ? U : never;
```

`getEffectiveTypeArguments` for `X1<infer U>` was doing all kinds of work to create the default type for `Output`, which eventually pulls on `T` who was instantiated with `U`, creating a circularity. But all this was in service of instantiating `HasM` which obviously doesn’t need to be instantiated. This simply lazifies getting those type argument types until/unless they’re needed (and also doesn’t bother filling in defaults if they’re never asked for).